cmd/traffic/cmd/manager/service.go

```diff
@@ -260,20 +267,165 @@ func (s *service) Remain(ctx context.Context, req *rpc.RemainRequest) (*empty.Em
 	ctx = managerutil.WithSessionInfo(ctx, session)
-	sessionID := tunnel.SessionID(session.GetSessionId())
 
+	sessionID := tunnel.SessionID(session.GetSessionId())
 	// There's no reason for the caller to wait for this removal to complete.
 	go s.state.RemoveSession(context.WithoutCancel(ctx), sessionID)
 	return &empty.Empty{}, nil
 }
 
+func (s *service) removeUnusedAgent(ctx context.Context, workloadKey *mutator.WorkloadKey) error {
+	maxIdleTime := managerutil.GetEnv(ctx).AgentMaxIdleTime
+	if maxIdleTime == 0 {
+		// default aka not set is 0, we don't ever remove agents, skip
+		return nil
+	}
+
+	activeIntercepts := s.state.CountActiveInterceptsForWorkload(workloadKey)
+	if activeIntercepts != 0 {
+		// don't remove if there are active intercepts
+		return nil
+	}
+
+	agentStateFileYAML := s.configWatcher.GetAgentStateYaml(ctx)
+	var agentStateFile AgentStateFile
+	err := yaml.Unmarshal(agentStateFileYAML, &agentStateFile)
+	if err != nil {
+		return fmt.Errorf("err in unmarshalling YAML: %w", err)
+	}
+
+	agentState, ok := agentStateFile.AgentStates[*workloadKey]
+	if !ok {
+		return fmt.Errorf("error in looking up workload key %s in agentStateFile, err: %w", workloadKey, err)
+	}
+	lastEngagementTime := agentState.LastEngagementTime
+	dlog.Tracef(ctx, "Last engagement time for agent %s: %s", *workloadKey, lastEngagementTime)
+	if lastEngagementTime.IsZero() {
+		// means it was never engaged
+		return nil
+	}
+	idleTime := time.Since(lastEngagementTime)
+	dlog.Tracef(ctx, "Idle time for agent %s: %s", *workloadKey, idleTime)
+	if idleTime > maxIdleTime {
+		// construct uninstall agents request
+		dlog.Infof(ctx, "Removing agent %s due to idle time %s exceeding max idle time %s", *workloadKey, idleTime, maxIdleTime)
+		ns := workloadKey.Namespace
+		mm := mutator.GetMap(ctx)
+		wl, err := k8sapi.GetWorkload(ctx, workloadKey.Name, ns, "")
+		if err != nil {
+			return status.Errorf(codes.NotFound, "Workload %s.%s not found", workloadKey.Name, ns)
+		}
+		mm.Delete(wl.GetName(), ns)
+		if err := mm.EvictPodsWithAgentConfig(ctx, wl); err != nil {
+			return status.Errorf(codes.Internal, "unable to delete agent for workload %s.%s: %v", wl.GetName(), ns, err)
+		}
+	}
+	return nil
+}
+
+func (s *service) UpdateLastEngagementTime(ctx context.Context, workloadKey *mutator.WorkloadKey) error {
+	// updates last engagement time IN MEMORY, this is persisted to the configmap by another goroutine that runs periodically
+	dlog.Tracef(ctx, "Logging workloadKey for last engagement time: %s", *workloadKey)
+
+	agentStateFileYAML := s.configWatcher.GetAgentStateYaml(ctx)
+	dlog.Tracef(ctx, "Logging agentStateFileYAML: %s", agentStateFileYAML)
+
+	var agentStateFile AgentStateFile
+	if string(agentStateFileYAML) != "" {
+		if s.state.CountActiveInterceptsForWorkload(workloadKey) == 0 {
+			// don't update last engagement time if there are no active intercepts and it is not the first time
+			return nil
```

Comment:
I think you need to check if the time for the workloadKey is the zero time and, if so, don't return here. How would you otherwise know that this "it is not the first time"? The `agentStateFileYAML` may contain other workloads.